crates/forge/src/link.rs

```diff
@@ -1,411 +1,130 @@
-    contracts::ArtifactContracts,
-    ArtifactId,
-};
+use foundry_compilers::{artifacts::Libraries, contracts::ArtifactContracts, ArtifactId};
+use semver::Version;
 use std::{
-    collections::{BTreeMap, HashMap},
-    fmt,
-    path::{Path, PathBuf},
+    collections::{HashMap, HashSet},
     str::FromStr,
 };
 
-/// Data passed to the post link handler of the linker for each linked artifact.
-#[derive(Debug)]
-pub struct PostLinkInput<'a, T, U> {
-    /// The fully linked bytecode of the artifact
-    pub contract: CompactContractBytecode,
-    /// All artifacts passed to the linker
-    pub known_contracts: &'a mut BTreeMap<ArtifactId, T>,
-    /// The ID of the artifact
-    pub id: ArtifactId,
-    /// Extra data passed to the handler, which can be used as a scratch space.
-    pub extra: &'a mut U,
-    /// Each dependency of the contract in their resolved form.
-    pub dependencies: Vec<ResolvedDependency>,
-}
-
-/// Dependencies for an artifact.
-#[derive(Debug)]
-struct ArtifactDependencies {
-    /// All references to dependencies in the artifact's unlinked bytecode.
-    dependencies: Vec<ArtifactDependency>,
-    /// The ID of the artifact
-    artifact_id: ArtifactId,
-}
-
-/// A dependency of an artifact.
-#[derive(Debug)]
-struct ArtifactDependency {
-    file: String,
-    key: String,
-    version: String,
-}
+fn find_artifact_id_by_library_path<'a>(
+    contracts: &'a ArtifactContracts,
+    file: &String,
+    name: &String,
+    version: Option<&Version>,
+) -> &'a ArtifactId {
+    for id in contracts.keys() {
+        if let Some(version) = version {
+            if id.version != *version {
+                continue;
+            }
+        }
+        // name is either {LibName} or {LibName}.{version}
+        if id.name.split('.').next().unwrap() != name {
+            continue;
+        }
 
-struct ArtifactCode {
-    code: CompactContractBytecode,
-    artifact_id: ArtifactId,
-}
+        if !(id.source.ends_with(file)) {
+            continue;
+        }
 
-impl std::fmt::Debug for ArtifactCode {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        self.artifact_id.fmt(f)
+        return id;
     }
-}
 
-#[derive(Debug)]
-struct AllArtifactsBySlug {
-    /// all artifacts grouped by identifier
-    inner: BTreeMap<String, ArtifactCode>,
+    panic!("artifact not found for library {file} {name}");
 }
 
-impl AllArtifactsBySlug {
-    /// Finds the code for the target of the artifact and the matching key.
-    fn find_code(&self, identifier: &String, version: &String) -> Option<CompactContractBytecode> {
-        trace!(target: "forge::link", identifier, "fetching artifact by identifier");
-        let code = self
-            .inner
-            .get(identifier)
-            .or(self.inner.get(&format!("{}.{}", identifier, version)))?;
-
-        Some(code.code.clone())
+pub fn collect_dependencies<'a>(
```

Comment:
Could we add docs explaining what this does? I think it's clear, but it's good for future reference.